File: integration-tests-java/src/main/java/io/micronaut/jimmer/java/it/repository/UserRoleRepository.java
package io.micronaut.jimmer.java.it.repository;

import io.micronaut.jimmer.java.it.config.Constant;
import io.micronaut.jimmer.java.it.entity.Fetchers;
import io.micronaut.jimmer.java.it.entity.Tables;
import io.micronaut.jimmer.java.it.entity.UserRole;
import io.micronaut.jimmer.java.it.entity.dto.UserRoleSpecification;
import io.micronaut.jimmer.repository.JRepository;
import io.micronaut.jimmer.repository.annotation.Repository;
import java.util.List;
import java.util.UUID;

@Repository(dataSourceName = Constant.DATASOURCE2)
public interface UserRoleRepository extends JRepository<UserRole, UUID> {

    default List<UserRole> find(UserRoleSpecification userRoleSpecification) {
        return sql().createQuery(Tables.USER_ROLE_TABLE)
                .where(userRoleSpecification)
                .select(Tables.USER_ROLE_TABLE.fetch(Fetchers.USER_ROLE_FETCHER.allScalarFields()))
                .execute();
    }

    UserRole findByUserId(String userId);

    UserRole findByRoleId(String roleId);

    UserRole findByUserIdAndRoleId(String userId, String roleId);
}

File: integration-tests-kotlin/src/main/kotlin/io/micronaut/jimmer/kotlin/it/repository/UserRoleRepository.kt
package io.micronaut.jimmer.kotlin.it.repository

import io.micronaut.jimmer.kotlin.it.entity.UserRole
import io.micronaut.jimmer.kotlin.it.entity.by
import io.micronaut.jimmer.kotlin.it.entity.dto.UserRoleSpecification
import io.micronaut.jimmer.repository.KRepository
import io.micronaut.jimmer.repository.annotation.Repository
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import java.util.UUID

@Repository(dataSourceName = "DB2")
interface UserRoleRepository : KRepository<UserRole, UUID> {
    fun find(userRoleSpecification: UserRoleSpecification): List<UserRole> =
        sql
            .createQuery(UserRole::class) {
                where(userRoleSpecification)
                select(table.fetch(newFetcher(UserRole::class).by { allScalarFields() }))
            }.execute()

    fun findByUserId(userId: String): UserRole

    fun findByRoleId(roleId: String): UserRole

    fun findByUserIdAndRoleId(
        userId: String,
        roleId: String,
    ): UserRole
}
